### PR TITLE
Implement three-legged authentication

### PR DESCRIPTION
**Pull request by AndrewLester**

### CODE REVIEW

**Comment by ErikBoesen** [comment]
Can you merge the upstream changes?

**Comment by andrewl1** [comment]
I thought I did, let me check.

File: schoolopy/authentication.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+        self.DOMAIN_ROOT = 'https://fccps.schoology.com'
+
+        self.user_id = user_id
+        self.consumer_key = consumer_key
+        self.consumer_secret = consumer_secret
+
+        self.oauth = requests_oauthlib.OAuth1Session(self.consumer_key, self.consumer_secret)
+
+        if user_id != None:
+            with open('auth.yml', 'r+') as f:
```
Comment (ErikBoesen) [inline schoolopy/authentication.py]: This sort of thing shouldn't be hardcoded; maybe figure out how to get this data through function parameters?

File: schoolopy/authentication.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+import yaml
+import requests_oauthlib
+import json
+from urllib import parse
+import requests
+
+class SchoologyAuth:
+    def __init__(self, consumer_key, consumer_secret, user_id=None):
+        self.API_ROOT = 'https://api.schoology.com/v1'
+        self.DOMAIN_ROOT = 'https://fccps.schoology.com'
```
Comment (ErikBoesen) [inline schoolopy/authentication.py]: This shouldn't be hardcoded.

File: schoolopy/authentication.py [hunk 3]
```diff
@@ -0,0 +1,135 @@
+import random
+import time
+import yaml
+import requests_oauthlib
+import json
+from urllib import parse
+import requests
+
+class SchoologyAuth:
```
Comment (ErikBoesen) [inline schoolopy/authentication.py]: I don't know if auth or authentication is the best thing to call this, nor if this is the best method of implementation. I think a good solution would be to merge this into the main `Schoology` class, and allow the number of auth legs to be specified when instantiating the class.

File: schoolopy/authentication.py (hunk 2 above)
Comment (ErikBoesen) [reply schoolopy/authentication.py]: It would only work for us.

File: schoolopy/authentication.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+    def _request_header(self):
+        header = {
+            'Authorization': self._oauth_header(),
+            'Accept': 'application/json',
+            'Host': 'api.schoology.com',
+            'Content-Type': 'application/json'
+        }
+        return header
+
+    def saveAccessTokens(self, uid, oauth_token, oauth_token_secret):
```
Comment (ErikBoesen) [inline schoolopy/authentication.py]: Small thing, but camelCase again

File: schoolopy/authentication.py [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+            fetch_response = self._fetch_token(request_token_url, self.oauth)
+
+
+            self.request_token = fetch_response.get('oauth_token')
+            self.request_token_secret = fetch_response.get('oauth_token_secret')
+            self.saveRequestTokens(self.request_token, self.request_token_secret)
+            base_authorization_url = self.DOMAIN_ROOT + '/oauth/authorize'
+
+            authorization_url = self.oauth.authorization_url(base_authorization_url)
+            print('Please go here and authorize,', authorization_url + r'&oauth_callback=https%3A%2F%2Ffccps.schoology.com')
```
Comment (ErikBoesen) [inline schoolopy/authentication.py]: Um.

File: schoolopy/authentication.py (hunk 2 above)
Comment (andrewl1) [reply schoolopy/authentication.py]: Yeah I just put that in to test. I think we can add a prefix thing for people to enter.

File: schoolopy/main.py [hunk 6]
```diff
@@ -147,11 +147,11 @@ def get_session(self):
 
     def get_me(self):
         """
-        Get data of the client using this method (Yourself)
+        Get data of the client using this method (Yourself).
 
         :return: User object obtained from API. (Of yourself)
         """
-        return User(self.get_user(self.get_self_user_info()['api_uid']))
+        return User(self.get_user(self.get_self_user_info().api_uid))
```
[the remark this answers is not in the document]
Comment (andrewl1) [reply schoolopy/main.py]: That endpoint only works if you have 3 leg enabled.  I'll make it switch if you have it on or off.

**Comment by ErikBoesen** [comment]
Another thing: I don't love how `auth.yml` is separate from the other example configuration.

**Comment by ErikBoesen** [comment]
You also need to add the additional dependencies to `setup.py`.

**Comment by andrewl1** [comment]
I think we could create a local server to callback the user to, if we want to know when they are authorized.

File: schoolopy/authentication.py [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+
+class AuthorizationError(Exception):
+    """Called when an attempt is made to initialize a Schoology instance with
+    an unauthorized SchoologyAuth instance"""
+    pass
+
+class SchoologyAuth:
+    def __init__(self, consumer_key, consumer_secret, domain_prefix='www', user_id=None, storage_file_name='example_config.yml'):
+        self.API_ROOT = 'https://api.schoology.com/v1'
+        self.DOMAIN_ROOT = 'https://%s.schoology.com' % domain_prefix
```
Comment (ErikBoesen) [inline schoolopy/authentication.py]: The domain root doesn't always match this pattern. Prefer specifying the entire domain.

File: schoolopy/authentication.py [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+            'Host': 'api.schoology.com',
+            'Content-Type': 'application/json'
+        }
+        return header
+
+    def save_access_tokens(self, uid, oauth_token, oauth_token_secret):
+        self.config[uid]['access_token'] = oauth_token
+        self.config[uid]['access_token_secret'] = oauth_token_secret
+        with open(self.storage_file, 'w') as f:
+            yaml.dump(self.config, f, default_flow_style=False)
```
Comment (ErikBoesen) [inline schoolopy/authentication.py]: Don't do this.

File: schoolopy/main.py [hunk 9]
```diff
@@ -56,7 +60,7 @@ def _post(self, path, data):
         :return: JSON response.
         """
         try:
-            return requests.post(self._ROOT + path, json=data, headers=self._request_header()).json()
+            return self.schoology_auth.oauth.post(url='%s%s?limit=%s' % (self._ROOT, path, self.limit), json=data, headers=self._request_header()).json()
```
Comment (ErikBoesen) [inline schoolopy/main.py]: I think I made the limit change upstream as well, you'll want to pull my updates.

**Comment by ErikBoesen** [comment]
Fam what up

**Comment by AndrewLester** [comment]
Finally made both three leg and 2 leg work with the new system.

File: schoolopy/main.py [hunk 10]
```diff
@@ -1977,15 +1984,15 @@ def send_message(self, subject, message, user_ids):
         :param user_ids: A list of user ids to send the message to.
         """
         recipients = ','.join([str(uid) for uid in user_ids])
-        return Message(self._post('messages', {'subject': subject, 'message': message.json(), 'recipient_ids': recipients}))
```
Comment (ErikBoesen) [inline schoolopy/main.py]: Don't remove this `.json()`

Comment (AndrewLester) [reply schoolopy/main.py]: In this case, the message is just text to put into it. I should probably make the parameter "message_text" now that I think about it.

File: schoolopy/main.py [hunk 11]
```diff
@@ -1968,24 +1955,24 @@ def get_message(self, message_id):
...
         Send a message to a user or users.
 
         :param subject: A string holding the subject of a message.
-        :param message: A string holding the body of a message.
+        :param message_text: A string holding the body of a message.
         :param user_ids: A list of user ids to send the message to.
         """
         recipients = ','.join([str(uid) for uid in user_ids])
-        return Message(self._post('messages', {'subject': subject, 'message': message.json(), 'recipient_ids': recipients}))
+        return Message(self._post('messages', {'subject': subject, 'message': message_text, 'recipient_ids': recipients}))
```
Comment (ErikBoesen) [inline schoolopy/main.py]: I'd call this `content` instead.

**Pull request merged**